### PR TITLE
Verify `PeerId` when dialing remote

### PR DESCRIPTION
**Pull request by driftluo**
Verify `PeerId` when dialing remote

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -1,21 +1,92 @@
...
+use secio::PeerId;
 use std::net::SocketAddr;
 
 /// Change multiaddr to socketaddr
 pub fn multiaddr_to_socketaddr(addr: &Multiaddr) -> Result<SocketAddr, ()> {
-    use multiaddr::Protocol;
+    let mut iter = addr.iter().peekable();
 
+    while iter.peek().is_some() {
+        match iter.peek() {
```
Comment (jjyr) [inline src/utils.rs]: Use a more robust way to extract ip and addr
``` rust
let mut ip = None;
let mut port = None;
for item in iter {
    match item {
        Protocol::Ip4(ip4) => ip = Some(IpAddr::V4(ip4)),
        Protocol::Ip6(ip6) => ip = Some(IpAddr::V6(ip6)),
        Protocol::Tcp(tcp_port) => port = Some(tcp_port)
    }
    if ip.is_some() && port.is_some() {
        break;
    }
}
```

File: src/service.rs [hunk 2]
```diff
@@ -516,7 +522,7 @@ where
         &mut self,
         mut handle: H,
         remote_pubkey: Option<PublicKey>,
-        address: Multiaddr,
+        mut address: Multiaddr,
```
Comment (jjyr) [inline src/service.rs]: Can we remove it?

File: src/utils.rs (hunk 1 above)
Comment (driftluo) [reply src/utils.rs]: I think the method of the iterator I used is not a problem, you can look at the test.

I don't want to have something like this:
```
/ip4/127.0.0.1/p2p/fdafdafdsafdsfds/tcp/1337
```

File: src/service.rs (hunk 2 above)
Comment (driftluo) [reply src/service.rs]: Because there is no peer id(p2p protocol) in the `multiaddr` that the listener gets, so here need to insert it here.

**Pull request merged**